builtin/credential/approle/secret_id.go

```diff
@@ -0,0 +1,500 @@
+package approle
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+// secretIDStorageEntry represents the information stored in storage
+// when a SecretID is created. The structure of the SecretID storage
+// entry is the same for all the types of SecretIDs generated.
+type secretIDStorageEntry struct {
+	// Accessor for the SecretID. It is a random UUID serving as
+	// a secondary index for the SecretID. This uniquely identifies
+	// the SecretID it belongs to, and hence can be used for listing
+	// and deleting SecretIDs. Accessors cannot be used as valid
+	// SecretIDs during login.
+	SecretIDAccessor string `json:"secret_id_accessor" structs:"secret_id_accessor" mapstructure:"secret_id_accessor"`
+
+	// Number of times this SecretID can be used to perform the login operation
+	SecretIDNumUses int `json:"secret_id_num_uses" structs:"secret_id_num_uses" mapstructure:"secret_id_num_uses"`
+
+	// Duration after which this SecretID should expire. This is
+	// croleed by the backend mount's max TTL value.
+	SecretIDTTL time.Duration `json:"secret_id_ttl" structs:"secret_id_ttl" mapstructure:"secret_id_ttl"`
+
+	// The time in UTC when the SecretID was created
```

Comment:
If this is ever output (e.g. during a read from an accessor) or some other client call please ensure that it is being output as an RFC value, e.g. `SecretIDNumUses.Format(time.RFC3339)`

When using `structs` to convert things into a map it's easy to forget to sanitize output. One of the reasons I've soured slightly on `structs`.

Same comment for ExpirationTime/LastUpdatedTime.